File: Session8/Java/DeleteInLinkedList.java
class DeleteInLinkedList
{
	Node head; 
	class Node
	{
		int data;
		Node next;
		Node(int d) 
		{
			data = d; 
			next = null; 
		}
	}

	public void insertionAtHead(int data)
	{

		Node new_Node = new Node(data);
		new_Node.next = head;
		head = new_Node;
	}

	public void deleteFromHead()
	{
		if(head==null)
		{
			return;
		}
		head = head.next;
	}

	public void deleteFromTail()
	{
		// TODO : delete a node when linkedlist length is 0 or 1
		Node prev = null;
		Node temp = head;

		while(temp.next!=null)
		{
			prev = temp;
			temp = temp.next;
		}

		prev.next = null;
	}

	public void printList()
	{
		Node temp = head;
		while (temp!=null)
		{
			System.out.print(temp.data+" "+"-->");
			temp = temp.next;
		}
	}

	public static void main(String[] args)
	{
		DeleteInLinkedList llist = new DeleteInLinkedList();

		llist.insertionAtHead(1);
		llist.insertionAtHead(2);
		llist.insertionAtHead(3);
		llist.insertionAtHead(4);
		llist.insertionAtHead(5);

		llist.deleteFromHead();
		llist.deleteFromTail();
		llist.deleteFromTail();

		llist.printList();
	}
}

File: Session8/C++/DeleteInLinkedList.cpp
#include <iostream>
using namespace std;

class Node{
public:
    int data;
    Node *next;

    Node(int d)
    {
        data = d;
        next = NULL;
    }
};

void insertionAtHead(Node*&head, int data)
{
    Node *new_Node = new Node(data);
    new_Node -> next = head;
    head = new_Node;
}

void deleteAtHead(Node*&head)
{
    if(head==NULL)
    {
        return;
    }
    head = head->next;
}

void deleteAtTail(Node*&head)
{
    // TODO : delete a node when linkedlist length is 0 or 1
    Node*prev = NULL;
    Node*temp = head;

    while(temp->next!=NULL)
    {
        prev = temp;
        temp = temp->next;
    }

    prev->next = NULL;
}

void print(Node*head)
{
    Node *temp = head;
    while(temp!=NULL)
    {
        cout<<temp->data<<"-->";
        temp = temp->next;
    }
    cout<<endl;
}

int main() {

    Node* head = NULL;  

    insertionAtHead(head, 1);
    insertionAtHead(head, 2);
    insertionAtHead(head, 3);
    insertionAtHead(head, 4);
    insertionAtHead(head, 5);

    deleteAtHead(head);
    deleteAtHead(head);
    deleteAtTail(head);

    print(head);
    return 0;
}

// 3->2
